crates/bindings-csharp/BSATN.Runtime/BSATN/Runtime.cs

```diff
@@ -1,476 +1,414 @@
-{
-    public UInt128 Read(BinaryReader reader)
-    {
-        var lower = reader.ReadUInt64();
-        var upper = reader.ReadUInt64();
-        return new(upper, lower);
+        public void Write(BinaryWriter writer, Element[] value) => enumerable.Write(writer, value);
+
+        public AlgebraicType GetAlgebraicType(ITypeRegistrar registrar) =>
+            enumerable.GetAlgebraicType(registrar);
     }
 
-    public void Write(BinaryWriter writer, UInt128 value)
+    // Special case for byte arrays that can be dealt with more efficiently.
+    public readonly struct ByteArray : IReadWrite<byte[]>
     {
-        writer.Write((ulong)value);
-        writer.Write((ulong)(value >> 64));
-    }
+        public static readonly ByteArray Instance = new();
 
-    public AlgebraicType GetAlgebraicType(ITypeRegistrar registrar) =>
-        new AlgebraicType.U128(default);
-}
-#endif
+        public byte[] Read(BinaryReader reader) => reader.ReadBytes(reader.ReadInt32());
 
-public readonly struct U256 : IReadWrite<SpacetimeDB.U256>
-{
-    public SpacetimeDB.U256 Read(BinaryReader reader)
-    {
-        var bsatn = new U128Stdb();
-        var lower = bsatn.Read(reader);
-        var upper = bsatn.Read(reader);
-        return new(upper, lower);
-    }
+        public void Write(BinaryWriter writer, byte[] value)
+        {
+            writer.Write(value.Length);
+            writer.Write(value);
+        }
 
-    public void Write(BinaryWriter writer, SpacetimeDB.U256 value)
-    {
-        var bsatn = new U128Stdb();
-        bsatn.Write(writer, value.Lower);
-        bsatn.Write(writer, value.Upper);
+        public AlgebraicType GetAlgebraicType(ITypeRegistrar registrar) =>
+            new AlgebraicType.Array(new AlgebraicType.U8(default));
     }
 
-    public AlgebraicType GetAlgebraicType(ITypeRegistrar registrar) =>
-        new AlgebraicType.U256(default);
-}
+    // This namespace contains implementations of IReadWrite for System types.
+    //
+    // The BSATN types are intentionally named 1:1 to corresponding System types,
+    // and the IReadWrite implementation uses the full type name even for primitives to make it clear.
+    //
+    // This allows to easily add more implementations for System types in the future (e.g. new collections)
+    // without having to touch the codegen at all, and without codegen having to hardcode any special types.
+    namespace SystemImpls
+    {
+        // This implementation is nearly identical to RefOption. The only difference is the constraint on T.
+        // Yes, this is dumb, but apparently you can't have *really* generic `T?` because,
+        // despite identical bodies, compiler will desugar it to very different
+        // types based on whether the constraint makes it a reference type or a value type.
+        public readonly struct Nullable<Inner, InnerRW> : IReadWrite<System.Nullable<Inner>>
+            where Inner : struct
+            where InnerRW : IReadWrite<Inner>, new()
```

Comment:
the RW can be constrained to a struct instead of new(), would that make the above comment fixed?